### PR TITLE
Fix #4929: Fix logic for moving early assignements in JS ctors.

### LINKED ISSUE #4929: Assertion failed: Trying to move a local VarDef after the super constructor call of a non-native JS class
When compiling my project using my own facades for magenta.ts and tone.js, I get following Scala.js assertion:

<details><summary>[error]   Trying to move a local VarDef after the super constructor call of a non-native JS class</summary>

<pre>

[error] Error while emitting Main.scala
[error] assertion failed:
[error]   Trying to move a local VarDef after the super constructor call of a non-native JS class at RangePosition(C:\Dev\scalaJsLocalVarDefAssert\src\main\scala\Main.scala, 207, 239, 277)
[error]      while compiling: C:\Dev\scalaJsLocalVarDefAssert\src\main\scala\Main.scala
[error]         during phase: jscode
[error]      library version: version 2.13.12
[error]     compiler version: version 2.13.12
[error]   reconstructed args: -bootclasspath C:\Users\Ondra\AppData\Local\Coursier\Cache\v1\https\repo1.maven.org\maven2\org\scala-lang\scala-library\2.13.12\scala-library-2.13.12.jar -classpath C:\Dev\scalaJsLocalVarDefAssert\target\scala-2.13\classes;C:\Users\Ondra\AppData\Local\Coursier\Cache\v1\https\repo1.maven.org\maven2\org\scala-js\scalajs-library_2.13\1.15.0\scalajs-library_2.13-1.15.0.jar;C:\Users\Ondra\AppData\Local\Coursier\Cache\v1\https\repo1.maven.org\maven2\org\scala-js\scalajs-scalalib_2.13\2.13.12%2B1.15.0\scalajs-scalalib_2.13-2.13.12%2B1.15.0.jar;C:\Users\Ondra\AppData\Local\Coursier\Cache\v1\https\repo1.maven.org\maven2\org\scala-js\scalajs-javalib\1.15.0\scalajs-javalib-1.15.0.jar -Xplugin:C:\Users\Ondra\AppData\Local\Coursier\Cache\v1\https\repo1.maven.org\maven2\org\scala-js\scalajs-compiler_2.13.12\1.15.0\scalajs-compiler_2.13.12-1.15.0.jar
[error]
[error]   last tree to typer: Select(Select(Ident(scala), scalajs), js)
[error]        tree position: line 6 of C:\Dev\scalaJsLocalVarDefAssert\src\main\scala\Main.scala
[error]             tree tpe: scala.scalajs.js.type
[error]               symbol: final package js in package scalajs
[error]    symbol definition: final package js (a ModuleSymbol)
[error]       symbol package: scala.scalajs
[error]        symbol owners: package js
[error]            call site: class Defined in package <empty>
[error]
[error] == Source file context for tree position ==
[error]
[error]      3
[error]      4 @js.native
[error]      5 @JSGlobal("Native")
[error]      6 class Native(output: js.Any = js.native, callbackObject: js.Any = js.native) extends js.Any
[error]      7
[error]      8 class Defined(callbackObject: js.Any) extends Native(callbackObject = callbackObject)
[error]      9
[error] one error found```

</pre>

</details> 

The minimal repro code (using Scala 2.13.18 and Scala.js 1.15.0):

_Note: I get no error when trying this with Scala 3.2.1 and Scala.js 1.15._

```scala
import scala.scalajs.js
import scala.scalajs.js.annotation.JSGlobal

@js.native
@JSGlobal("Native")
class Native(output: js.Any = js.native, callbackObject: js.Any = js.native) extends js.Any

class Defined(callbackObject: js.Any) extends Native(callbackObject = callbackObject)
```

I guess there is probably some error in my code or in the facade, but the assertion is not very helpful.

Online repro: https://scribble.ninja/u/OndrejSpanel/crxnynclgqgzlfdcltblxjpofdqy

### PR DESCRIPTION
**Pull request by sjrd**
Previously, we moved all statements in the constructors after the super constructor call. However, it turns out that there are statements that must be kept before, notably local `val`s generated for default arguments to the super constructor.

We now keep statements where they are by default. We only move statements of the form `C.this.field = ident;`, which are the only ones that require access to `this`.

This requires only a little bit of special-casing for outer pointer null checks. Fortunately, we already had some logic to identify and decompose those, which we reuse here.

---

@gzm0 I started from your partial solution from https://github.com/scala-js/scala-js/issues/4929#issuecomment-1913659557, except the "it needs to be a bit more principled" turned out to require a complete rewrite of the surrounding logic. :sweat_smile:

### CODE REVIEW

File: test-suite/js/src/test/require-scala2/org/scalajs/testsuite/jsinterop/NonNativeJSTypeTestScala2.scala [hunk 1]
```diff
@@ -46,4 +65,20 @@ object NonNativeJSTypeTestScala2 {
 
   class SomeValueClass(val i: Int) extends AnyVal
 
+  object ConstructorSuperCallWithDefaultParamsAndEarlyInitializers {
+    val sideEffects = mutable.ListBuffer.empty[String]
+
+    class Parent(parentParam1: Any = "param1", parentParam2: Any = "param2") extends js.Object {
```
Comment (gzm0) [inline test-suite/js/src/test/require-scala2/org/scalajs/testsuite/jsinterop/NonNativeJSTypeTestScala2.scala:71]: Use a default parameter that depends on another parameter? (IIRC this is allowed with multiple argument lists).

File: test-suite/js/src/test/scala/org/scalajs/testsuite/jsinterop/NonNativeJSTypeTest.scala [hunk 2]
```diff
@@ -2040,6 +2057,19 @@ object NonNativeJSTypeTest {
     def this(x: Int, y: Int) = this(x)(y.toString, js.undefined)
   }
 
+  object ConstructorSuperCallWithDefaultParams {
+    val sideEffects = mutable.ListBuffer.empty[String]
+
+    class Parent(parentParam1: Any = "param1", parentParam2: Any = "param2") extends js.Object {
```
Comment (gzm0) [inline test-suite/js/src/test/scala/org/scalajs/testsuite/jsinterop/NonNativeJSTypeTest.scala:2063]: Same here: Make the default parameter dependent on another parameter?

**Review by gzm0 — CHANGES_REQUESTED** [review]
Just two additional suggestions for tests.

File: test-suite/js/src/test/require-scala2/org/scalajs/testsuite/jsinterop/NonNativeJSTypeTestScala2.scala (hunk 1 above)
Comment (sjrd) [reply test-suite/js/src/test/require-scala2/org/scalajs/testsuite/jsinterop/NonNativeJSTypeTestScala2.scala]: Good point! Done.

**Review by gzm0 — APPROVED** [review]

**Pull request merged**